### ISSUE TITLE
Directory.Delete(string, bool) throws StackOverflowException

### ISSUE THREAD

**Issue by tahirkilic** [dotnet/runtime#84344]
### Description

Directory.Delete(string, bool) throws a StackOverflowException if the nesting depth of subdirectories is huge (in my case 14000).

### Reproduction Steps

- Create nested subdirectories like below:
```csharp
    DirectoryInfo topMost = new DirectoryInfo("C:\\temp");
    int depth = 14000;

    DirectoryInfo subDir = topMost;
    for (int i = 0; i < depth; i++)
    {
        subDir = subDir.CreateSubdirectory("a");
    }
```
- Then try to delete the top most directory 
```csharp
Directory.Delete(topMost.FullName, true);
```

### Expected behavior

Topmost directory and its all subdirectories are deleted successfully.

### Actual behavior

A StackOverflowException is thrown with the following stacktrace:
```csharp
Stack overflow.
   at System.String.Concat(System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>)
   at System.IO.Path.JoinInternal(System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>)
   at System.IO.Path.CombineInternal(System.String, System.String)
   at System.IO.Path.Combine(System.String, System.String)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   at System.IO.FileSystem.RemoveDirectoryRecursive(System.String, WIN32_FIND_DATA ByRef, Boolean)
   ....
   ....
```

### Configuration

**dotnet version:** 6.0.301

**operating system:** Windows 10 version 21H2

**Comment by GoldenretriverYT** [comment]
I dont think this is a bug as you have specified `recursive: true` and StackOverflows are a risk of recursive actions.

**Comment by aevitas** [comment]
> I suppose this method could probe for stack space in order to fail more usefully.

How would it do that? Could a method figure out that it itself had been called N times, and if so how would it know when to terminate execution? Would you have the method walk the stack every time it's called, and if so, wouldn't that be a huge performance impact on non-academic usage of the method? I'm genuinely curious how this could be done.

**Comment by eriawan** [comment]
@adamsitnik 

Is this issue already taken or is currently the work? If it is not, I would like to try to take this issue by assigning this issue to me.

I have an idea to safely recurse the folders without wasting stack, using a bit of fixed point combinator. 🙂 
I may want to promote/propose a new class to feature that later.

**Comment by danmoseley** [comment]
@eriawan I will assign to you. thanks for working on this!

**Comment by danmoseley** [comment]
> I have an idea to safely recurse the folders without wasting stack, using a bit of fixed point combinator.

To fix this specific bug, it is probably only needed to call RuntimeHelpers. I suggest to make that limited fix in its own PR first. If you want to avoid recursion entirely, that would be a larger change, and would need some perf measurements to demonstrate it was worthwhile.

**Comment by eriawan** [comment]
@danmoseley 

> To fix this specific bug, it is probably only needed to call RuntimeHelpers. I suggest to make that limited fix in its own PR first. If you want to avoid recursion entirely, that would be a larger change, and would need some perf measurements to demonstrate it was worthwhile.

Thanks for the suggestions above! About the `RuntimeHelpers.TryEnsureSufficientExecutionStack` and `RuntimeHelpers.EnsureSufficientExecutionStack`, I have tried to read at MS Learn docs, and there is no sample code on using that. Is there any sample code/usage in the dotnet/runtime repo?

**Comment by danmoseley** [comment]
There should be usage in this repo: I can't check right now.

**Comment by jkotas** [comment]
This is fairly typical usage pattern: https://github.com/dotnet/runtime/blob/5b8437d9f9e410bf672bb62d5371fdeb8931e5f1/src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/StackGuard.cs#L27-L28

**Comment by eriawan** [comment]
@jkotas 
Thanks! I think I need some time to try that `RuntimeHelpers.EnsureSufficientExecutionStack` locally and try to apply some use cases before I submit my PR.

Hopefully this weekend I'm going to submit draft PR the earliest, otherwise next week.

**Comment by danmoseley** [comment]
Sounds good. Please (of course) include a test that fails before the fix and passes after. Basically the repro case above but probably inside TestDirectory [similar to this](https://github.com/dotnet/runtime/blob/3747edb1812106f0b88a42b724c3e44b69dc79c4/src/libraries/System.IO.FileSystem/tests/DirectoryInfo/CreateSubdirectory.cs#L42)

**Comment by eriawan** [comment]
> Sounds good. Please (of course) include a test that fails before the fix and passes after. Basically the repro case above but probably inside TestDirectory [similar to this](https://github.com/dotnet/runtime/blob/3747edb1812106f0b88a42b724c3e44b69dc79c4/src/libraries/System.IO.FileSystem/tests/DirectoryInfo/CreateSubdirectory.cs#L42)

Aha. Ok I'll create a new unit test to test this recursive scenario of Directory.Delete in this folder, will add to my PR.
